File: Binary_Search_Tree_Iterator_173.java
import java.util.Stack;

class Treenode {
      int val;
      Treenode left;
      Treenode right;
      Treenode() {}
      Treenode(int val) { this.val = val; }
      Treenode(int val, Treenode left, Treenode right) {
          this.val = val;
          this.left = left;
          this.right = right;
     }
 }


class Binary_Search_Tree_Iterator_173 {
    Stack<Treenode> st = new Stack<>();
    
    public Binary_Search_Tree_Iterator_173(Treenode root) {
        dfs(root);
    }
    
    /** @return the next smallest number */
    public int next() {
        Treenode poppedNode = st.pop();
        if(poppedNode.right != null)
            dfs(poppedNode.right);
        return poppedNode.val;
    }
    
    /** @return whether we have a next smallest number */
    public boolean hasNext() {
        return (!st.isEmpty());
    }
    
    private void dfs(Treenode root)
    {
        while(root != null)
        {
            st.push(root);
            root = root.left;
        }
    }
}

File: binary-search-tree-iterator_173.py
#Approach 2: 
#Time Complexity O(1) for two methods there will be pushes and sometimes there wont be, 
#also if we assume total N elements are pushed then N next calls must be there so on an average (N/N) ~ O(1)
#Space Complexity O(H) height of tree, Because max we are appending all the left nodes of root at any point not all N elements
#It successfully runs on the leetcode
class BSTIterator:
    s = []
    def __init__(self, root: Optional[TreeNode]):
        self.inorder(root)
        
    def inorder(self, root):
        while root!=None:
            self.s.append(root)
            root= root.left
        
        
    def next(self) -> int:
        if self.s: 
            node = self.s.pop()
            if node.right:
                self.inorder(node.right)
            return node.val        
        

    def hasNext(self) -> bool:
        return len(self.s)!=0




#Approach 1: storing traversal values in list and calling methods in recursive inorder traversal
#Time Complexity O(1) for two methods
#Space Complexity O(N) List
#It successfully runs on the leetcode
class BSTIterator:
    l=[]

    def __init__(self, root: Optional[TreeNode]):
        self.inorder(root)
        
    
    def inorder(self, root):
        if root == None:
            return 
        self.inorder(root.left)
        self.l.append(root.val)
        self.inorder(root.right)
        return self.l

    def next(self) -> int:
        if self.l:
            val = self.l.pop(0)
            return val
        else:
            return None
        

    def hasNext(self) -> bool:
        return len(self.l) > 0